documentation/src/main/asciidoc/topics/query.adoc

```diff
@@ -18,480 +18,570 @@ Apart from indexed queries, {brandname} can run queries over non-indexed or
-@Indexed
-public class Book {
-   @Field String title;
-   @Field String description;
-   @Field @DateBridge(resolution=Resolution.YEAR) Date publicationYear;
-   @IndexedEmbedded Set<Author> authors = new HashSet<Author>();
-}
+=== Specifying indexed Entities
 
-----
+It is recommended to declare the indexed types, as they will be mandatory in the next {brandname} version.
+This can be done via xml:
 
-[source,java]
-.Author.java
+[source,xml,options="nowrap",subs=attributes+]
 ----
-
-public class Author {
-   @Field String name;
-   @Field String surname;
-   // hashCode() and equals() omitted
-}
-
+include::config_examples/indexed_entities.xml[]
 ----
 
-Now assuming we stored several _Book_ instances in our {brandname} _Cache_ , we can search them for any matching field as in the following example.
-
-Using a Lucene Query:
+or programmatically:
 
 [source,java]
 ----
-// get the search manager from the cache:
-SearchManager searchManager = org.infinispan.query.Search.getSearchManager(cache);
-
-// create any standard Lucene query, via Lucene's QueryParser or any other means:
-org.apache.lucene.search.Query fullTextQuery = //any Apache Lucene Query
-
-// convert the Lucene query to a CacheQuery:
-CacheQuery cacheQuery = searchManager.getQuery( fullTextQuery );
-
-// get the results:
-List<Object> found = cacheQuery.list();
+ cacheCfg.indexing()
+       .addIndexedEntity(Car.class)
+       .addIndexedEntity(Truck.class)
 
 ----
 
-A Lucene Query is often created by parsing a query in text format such as "title:infinispan AND authors.name:sanne", or by using the query builder provided by Hibernate Search.
+When the cache is storing protobuf, the indexed types should be the _Message_ declared in the protobuf schema.
+For example, for the schema below:
 
-[source,java]
+[source,proto]
 ----
-// get the search manager from the cache:
-SearchManager searchManager = org.infinispan.query.Search.getSearchManager( cache );
-
-// you could make the queries via Lucene APIs, or use some helpers:
-QueryBuilder queryBuilder = searchManager.buildQueryBuilderForClass(Book.class).get();
-
-// the queryBuilder has a nice fluent API which guides you through all options.
-// this has some knowledge about your object, for example which Analyzers
-// need to be applied, but the output is a fairly standard Lucene Query.
-org.apache.lucene.search.Query luceneQuery = queryBuilder.phrase()
-                  .onField("description")
-                  .andField("title")
-                  .sentence("a book on highly scalable query engines")
-                  .createQuery();
-
-// the query API itself accepts any Lucene Query, and on top of that
-// you can restrict the result to selected class types:
-CacheQuery query = searchManager.getQuery(luceneQuery, Book.class);
-
-// and there are your results!
-List objectList = query.list();
-
-for (Object book : objectList) {
-      System.out.println(book);
-}
-
+include::config_examples/library.proto[]
```

Comment:
we can do this in the doc PR but I think [library.proto](https://github.com/infinispan/infinispan/blob/master/documentation/src/main/asciidoc/topics/config_examples/library.proto) should be renamed to book.proto.

Reply:
yes, let's do it later, as this file is reused elsewhere